Python/libraries/recognizers-date-time/recognizers_date_time/date_time/base_date.py

```diff
@@ -0,0 +1,768 @@
+    @abstractmethod
+    def date_unit_regex(self) -> Pattern:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def for_the_regex(self) -> Pattern:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def week_day_and_day_of_month_regex(self) -> Pattern:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def relative_month_regex(self) -> Pattern:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def week_day_regex(self) -> Pattern:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def day_of_week(self) -> Dict[str, int]:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def ordinal_extractor(self) -> BaseNumberExtractor:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def integer_extractor(self) -> BaseNumberExtractor:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def number_parser(self) -> BaseNumberParser:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def duration_extractor(self) -> DateTimeExtractor:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def utility_configuration(self) -> DateTimeUtilityConfiguration:
+        raise NotImplementedError
+
+class BaseDateExtractor(DateTimeExtractor):
+    @property
+    def extractor_type_name(self) -> str:
+        return Constants.SYS_DATETIME_DATE
+
+    def __init__(self, config: DateExtractorConfiguration):
+        self.config = config
+
+    def extract(self, source: str, reference: datetime = None) -> List[ExtractResult]:
+        if reference is None:
+            reference = datetime.now()
+
+        tokens = self.basic_regex_match(source)
+        tokens.extend(self.implicit_date(source))
+        tokens.extend(self.number_with_month(source, reference))
+        tokens.extend(self.duration_with_before_and_after(source, reference))
+
+        result = merge_all_tokens(tokens, source, self.extractor_type_name)
+        return result
+
+    def basic_regex_match(self, source: str) -> List[Token]:
+        ret: List[Token] = list()
+        for regexp in self.config.date_regex_list:
+            ret.extend(get_tokens_from_regex(regexp, source))
+
+        return ret
+
+    def implicit_date(self, source: str) -> List[Token]:
+        ret: List[Token] = list()
+        for regexp in self.config.implicit_date_list:
+            ret.extend(get_tokens_from_regex(regexp, source))
+
+        return ret
+
+    def number_with_month(self, source: str, reference: datetime) -> List[Token]:
+        ret: List[Token] = list()
+        extract_results = self.config.ordinal_extractor.extract(source)
+        extract_results.extend(self.config.integer_extractor.extract(source))
+
+        for result in extract_results:
+            num = int(self.config.number_parser.parse(result).value)
+            if num < 1 or num > 31:
```

Comment:
Nit: always add blank line before if/for/etc to make code a little easier to read.

Reply:
blank lines added